src/errors.rs

```diff
@@ -0,0 +1,180 @@
+//! Error definitions
+
+pub mod i2c {
+    /// I2C error
+    pub trait Error: core::fmt::Debug {
+        /// Convert error to a generic I2C error kind
+        ///
+        /// By using this method I2C errors freely defined by HAL implementations
+        /// can be converted to a set of generic I2C errors upon which generic
+        /// code can act.
+        fn kind(&self) -> ErrorKind;
+    }
+
+    /// I2C error kind
+    ///
+    /// This represents a common set of I2C operation errors. HAL implementations are
+    /// free to define more specific or additional error types. However, by providing
+    /// a mapping to these common I2C errors, generic code can still react to them.
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    #[non_exhaustive]
+    pub enum ErrorKind {
+        /// An unspecific bus error occurred
+        Bus,
+        /// The arbitration was lost, e.g. electrical problems with the clock signal
+        ArbitrationLoss,
+        /// A bus operation was not acknowledged, e.g. due to the addressed device not being available on
+        /// the bus or the device not being ready to process requests at the moment
+        NoAcknowledge,
+        /// The peripheral receive buffer was overrun
+        Overrun,
+        /// The peripheral send buffer ran out of data
+        Underrun,
```

Comment:
I think no current HAL uses this error kind so far: https://github.com/rust-embedded/embedded-hal/issues/229#issuecomment-647428967 But I assume there's still value if the error is standardized, in case it could be used in the future.

(However, in that case, would SPI and serial need an `Underrun` error as well?)

Reply:
Good point. As I said I mostly got the variants from `embedded-error`. I am fine with removing the `Underrun` variant since it seems unclear it can be detected at all.
The `ErrorKind`s are marked as `#[non_exahustive]` so adding it in the future is not a breaking change.